go/apps/mixnet/proxy.go

```diff
@@ -39,10 +41,57 @@ func NewProxyContext(path string) (p *ProxyContext, err error) {
 }
 
 // DialRouter connects anonymously to a remote Tao-delegated mixnet router.
-func (p ProxyContext) DialRouter(network, addr string) (net.Conn, error) {
+func (p *ProxyContext) DialRouter(network, addr string) (net.Conn, error) {
 	c, err := tao.Dial(network, addr, p.domain.Guard, p.domain.Keys.VerifyingKey, nil)
 	if err != nil {
 		return nil, err
 	}
 	return &Conn{c}, nil
 }
+
+// Serialize and pad a directive to the length of a cell and send it to the
+// router. A directive is signaled to the receiver by the first byte of the
+// cell. If the serialized protocol buffer doesn't fit into a cell, c.Write()
+// will throw an error.
+func (p *ProxyContext) SendDirective(c net.Conn, d *Directive) (n int, err error) {
+	db, err := proto.Marshal(d)
+	if err != nil {
+		return 0, err
+	}
+	return c.Write(padCell(append([]byte{dirCell}, db...)))
+}
+
+// CreateCircuit directs the router to construct a circuit to a particular
+// destination over the mixnet.
+func (p *ProxyContext) CreateCircuit(c net.Conn, circuitAddrs []string) (n int, err error) {
+	var d Directive
+	d.Type = DirectiveType_CREATE_CIRCUIT.Enum()
+	d.Addrs = circuitAddrs
+	return p.SendDirective(c, &d)
+}
+
+// SendMessage directs the router to relay a message over the already constructed
+// circuit. A message is signaled to the reecevier by the first byte of the first
+// cell. The next 8 bytes encode the total number of bytes in the message.
+func (p *ProxyContext) SendMessage(c net.Conn, msg []byte) (n int, err error) {
+	msgBytes := len(msg)
+	cell := make([]byte, CellBytes)
+	cell[0] = msgCell
+
+	// TODO(cjpatton) How to deal with endianness discrepancies?
```

Comment:
binary.PutUvarint has a well-defined byte order (from least-significant to most-significant), so I think if you use that (instead of binary.BigEndian.PutUint64), you should be ok.